### PR TITLE
Merge 3.1.x fixes into 4.0

### LINKED ISSUE #519: Merge 3.1.x fixes into 4.0
We have #478 and #484 which are fixed only on 3.1.0 branch, and a pending PR #516 on 3.1.0 which also want to travel into master.

We should merge them into master too.

### PR DESCRIPTION
**Pull request by iloveeclipse**
This fixes #519
----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code
- [ ] Added an entry into `gradlePlugin/CHANGELOG.md` if you have changed Gradle plugin code

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,18 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 
 ## Unreleased - 2017-??-??
```
Comment (KengoTODA) [inline CHANGELOG.md]: If we merge this PR, reader cannot understand which version will release following `Unreleased` fixes.
I think it's better to merge not `release-3.1` branch but the tagged commit as `3.1.1`.

**Comment by iloveeclipse** [comment]
> If we merge this PR, reader cannot understand which version will release following 

I don't think so. As far as neither 4.0.0 nor 3.1.2 is released, the release notes will show "Unreleased".
As soon as 3.1.2 is released, it will show "fixed in 3.1.2", and that note we simply merge it to 4.0.0 again (if there will be more fixes) or just update release notes on 4.0.0, so it still will have the right entry.

In jgit/egit we do this pretty often, and nobody complained so far. The changes are merged to the maintenance first, then to master, and whoever releases first, contains the right changelog.

**Comment by KengoTODA** [comment]
Wmm I am still afraid that human can easily mistake about which line is for 3.1 or 4.0. At lease I can mix them in CHANGELOG.md.

If you don't care, I want to release 3.1.2 first. If OK I will propose PR to do it.

**Comment by iloveeclipse** [comment]
> Wmm I am still afraid that human can easily mistake about which line is for 3.1 or 4.0. At lease I can mix them in CHANGELOG.md.

- The *released* CHANGELOG will do NOT have the "Unreleased" line.
- The *unreleased* CHANGELOG is for developers only, and for developers, everything which is not yet released but in the current branch is under "Unreleased".

It makes no sense to bump a maintenance  release for every commit we merge to master.

If we do a release of the maintenance, all new entries will get a release and next merge commit to master will move the entries from "Unreleased" to the appropriate maintenance version.

**Comment by iloveeclipse** [comment]
@KengoTODA : we should proceed here. Do you still have objections?

**Pull request merged**